lib/streamlit/runtime/caching/cache_utils.py

```diff
@@ -239,10 +241,27 @@ def _get_or_create_cached_value(
         with contextlib.suppress(CacheKeyNotFoundError):
             cached_result = cache.read_result(value_key)
             return self._handle_cache_hit(cached_result)
-        return self._handle_cache_miss(cache, value_key, func_args, func_kwargs)
+
+        # only show spinner if there is a message to show and always only for the
+        # outermost cache function if cache functions are nested, because the outermost
+        # function has to wait for the inner functions anyways. This avoids surprising
+        # users with slowdowned apps in case the inner functions are called very often,
+        # which would lead to a ton of (empty/spinner) proto messages that will make the
+        # app slow (see https://github.com/streamlit/streamlit/issues/9951). This is
+        # basically like auto-setting "show_spinner=False" on the @st.cache decorators
+        # on behalf of the user.
+        is_nested_cache_function = in_cached_function.get()
+        spinner_or_no_context = (
+            spinner(spinner_message, _cache=False)
```

Comment:
Why False?

Reply:
oh oops, great catch! This is a leftover from some testing. I will set it back to `True` again.

With regards to unit tests: yep, there is a TODO in the PR description. I first wanted to get feedback about the fix itself before adding unit tests 🙂 Will add them now that the approach is approved!